docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -20,401 +51,341 @@ The scenarios of defining placement rules in SQL include:
 
 ```sql
-ALTER TABLE test
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=bj]" ROLE=follower REPLICAS=2,
-	ALTER PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=voter REPLICAS=2;
+ALTER TABLE t1
+	LEARNER_CONSTRAINTS="[+engine=tiflash]" LEARNERS=1;
 ```
 
-There are 4 replicas for table `test`, 2 of which are in `sh` and 2 are in `bj`.  Leader can only be placed on `sh`.
-
-`ROLE` in the statement is equivalent to field `role` in PD placement rule configuration.
-
-### Replicas configuration
+The only way to judge whether it’s adding a TiFlash replica is to check the label. If it contains `engine=tiflash`, then it’s adding or removing a TiFlash replica. This logic is conventional in PD for now.
 
-`REPLICAS` in the statement indicates the replica count of the specified role.
+#### Specifying role count
 
-Rules defined on `leader` can omit `REPLICAS`, because the count of leader is always 1.
+The roles `FOLLOWERS`, `LEARNERS` and `VOTERS` also support an optional count. For example:
 
-When all the replica counts are specified in the `CONSTRAINTS` option, `REPLICAS` can also be omitted. For example, `CONSTRAINTS="{+zone=bj:2,+zone=sh:1}", ROLE=voter` indicates that the `REPLICAS` is 3.
+```sql
+CREATE PLACEMENT POLICY `standardplacement1` PRIMARY REGION="us-east-1" REGIONS="us-east-1,us-east-2" FOLLOWERS=4;
+CREATE PLACEMENT POLICY `standardplacement2` LEADER_CONSTRAINTS="[+region=us-east-1]"  FOLLOWERS_CONSTRAINTS="[+region=us-east-1,+region=us-east-2]" FOLLOWERS=4;
+```
 
-When both `REPLICAS` and `count` in `CONSTRAINTS` are specified, it indicates that the other replicas can be placed anywhere. For example, in the case `CONSTRAINTS="{+zone=bj:2,+zone=sh:1}", ROLE=voter, REPLICAS=4`, 2 replicas are in `bj` and 1 in `sh`, and the last replica can be anywhere, including `bj` and `sh`.
+If the constraints is specified as a dictionary (e.g. `{"+region=us-east-1":1}`), and the count is smaller than the minimum number required to satisfy the constraint rule, then an error should be returned.
+Constraints do not automatically increase the count of the role, so if a larger number than the default is required it must be specified:
 
-When the `CONSTRAINTS` option doesn't contain `count`, `REPLICAS` must be specified. For example, `CONSTRAINTS="[+zone=bj]" ROLE=follower` is vague, as the count of `follower` can not be inferred.
+```sql
+FOLLOWER_CONSTRAINTS="{+region=us-east-1:1,-region=us-east-2:2}" FOLLOWERS=2 // error, requires 3
+FOLLOWER_CONSTRAINTS="{+region=us-east-1:1,-region=us-east-2:2}" // default is 2, so also error
+```
 
-`REPLICAS` in the statement is equivalent to field `count` in PD placement rule configuration.
+This changes an earlier proposal where the minimum count could be inferred by the constraints. The motivation for this change is to reduce the risk of misconfiguration in the event that rules are complex. See also "Policy Validation" below.
```

Comment:
@xhebox Can you review this proposed change? I don't believe we have discussed it, but when reading through I realized this was another ambigous case that could lead to misconfiguration.

Reply:
Actually, I would suggest ban using dict and replicas... It is not intuitive for me. Maybe replace `FOLLOWER_CONSTRAINTS="{+region=us-east-1:1,-region=us-east-2:2}" FOLLOWERS=6` with `FOLLOWER_CONSTRAINTS="{+region=us-east-1:1,-region=us-east-2:2, *:3}"` or so.

Reply:
Makes sense to me. I'm just a little hung up on the `*` character. Is there an alternative syntax we can use? Maybe `any:3`?

Reply:
No, I did not come up with any good. "any", "others", "free", or maybe just "replicas/count"?

Reply:
count/replicas is ambiguous because it doesn't indicate if it is inclusive or exclusive of the previously defined ones. I will go with `any`, and we can change it if needed later.

Reply:
Fixed in https://github.com/pingcap/tidb/pull/26221/commits/9ed4000bfd9ba578eff5d43aaaccf53c6300bcff